lib/internal/streams/compose.js

```diff
@@ -0,0 +1,234 @@
+'use strict';
+
+const {
+  Symbol
+} = primordials;
+const pipeline = require('internal/streams/pipeline');
+const Duplex = require('internal/streams/duplex');
+const { createDeferredPromise } = require('internal/util');
+const PassThrough = require('internal/streams/passthrough');
+const eos = require('internal/streams/end-of-stream');
+const { destroyer } = require('internal/streams/destroy');
+const assert = require('internal/assert');
+const from = require('internal/streams/from');
+const {
+  isStream,
+  isIterable
+} = require('internal/streams/utils');
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_MISSING_ARGS,
+  },
+} = require('internal/errors');
+const kDidRead = Symbol('kDidRead');
+class ComposeDuplex extends Duplex {
+  constructor(options) {
+    super(options);
+
+    // https://github.com/nodejs/node/pull/34385
+
+    if (options?.readable === false) {
+      this._readableState.readable = false;
+      this._readableState.ended = true;
+      this._readableState.endEmitted = true;
+    }
+
+    if (options?.writable === false) {
+      this._writableState.writable = false;
+      this._writableState.ending = true;
+      this._writableState.ended = true;
+      this._writableState.finished = true;
+    }
+
+    this[kDidRead] = false;
+  }
+
+  push(chunk, encoding) {
+    const ret = Duplex.prototype.push.call(this, chunk, encoding);
+    if (chunk != null) {
+      this[kDidRead] = true;
+    } else if (!this[kDidRead]) {
+      this.resume();
+    }
+    return ret;
+  }
+}
```

Comment:
@mcollina: A bit of a crazy idea... it would be possible here to implement a stream v4 that can **only** be used through `compose/pipeline/finished`, i.e. all the methods `push, resume, pause, write, end, on` don't do anything, only the property accessor are available. So `compose` returns streams that can only be used through the stream helpers.

Reply:
We could also implement "lock", similar to web streams and possibly provide better compat/support for web streams. @jasnell

Reply:
Something like the following public interface:

```js
class StreamObj extends EE {
  abort();
  get locked();
}
```

Reply:
Maybe a bit to crazy... I'll revisit this later.